src/services/completions.ts

```diff
@@ -1182,7 +1182,7 @@ namespace ts.Completions {
         function filterGlobalCompletion(symbols: Symbol[]): void {
             const isTypeOnly = isTypeOnlyCompletion();
             const allowTypes = isTypeOnly || !isContextTokenValueLocation(contextToken) && isPossiblyTypeArgumentPosition(contextToken, sourceFile, typeChecker);
-            if (isTypeOnly) {
+            if (allowTypes) {
                 keywordFilters = isTypeAssertion()
                     ? KeywordCompletionFilters.TypeAssertionKeywords
                     : KeywordCompletionFilters.TypeKeywords;

```

Comment:
note that this means we no longer offer *other* keywords in completions; `new, this, typeof` are the useful ones I saw from scanning the list. Probably `this` and `typeof` should be added to the list `typeKeywords`. Suggesting `new` isn't very useful since it only saves 3 characters (or less).

For example,

```ts
return x < this.property;
```

will not suggest `this` any more.

Reply:
> Probably `this` and `typeof` should be added to the list `typeKeywords`.

Agreed. That seems objectively correct, since those _are_ valid in a type position.

But why wouldn’t we show both type and non-type keywords here, given the uncertainty? Related questions:

1. Are we trying to avoid semantic work here? It seems possible to differentiate between an `x` that can’t possibly accept type parameters (so we’re definitely not in a type position) vs. an `x` that’s generic (so we may or may not be).
2. If we are trying to avoid semantic work, would it be reasonable to use whitespace as a heuristic? `Map <string, number>` seems extraordinarily uncommon, and `x<3` fairly uncommon too.

I’m not sure whether these are practical or not, but I don’t think they’re necessary for this PR—seems like an improvement already. Just food for thought. (I do think adding `this` and `typeof` now is probably worth it, though.)

Reply:
Hmm. Both the simple "show all" idea and "be smarter about possible type arguments" idea are good.

For the second, isPossiblyTypeArgumentPosition *already* calls `getTypeAtLocation`, so it's just a matter of querying the returned value.

I added `this` and `typeof` as types and it broke find all refs of global `this` horribly. I haven't looked at why, but it means that I'll probably keep them where they are for now.

Reply:
Yep, `isPossiblyTypeArgumentPosition` already is smarter about possible type arguments. My tests just don't show it.

Reply:
In any case, my original example, `x < this.property`, is wrong -- it *does* suggest `this`. It just doesn't when the identifier to the left of `<` is a generic signature. 

So I think making `this` and `typeof` type keywords is best fixed in a separate PR, since it requires more knowledge of services than I have now, and only affects things that are genuinely type contexts. I filed a bug #32486 to track it.

Reply:
Ah, that sounds better. I was a little concerned that you’d be missing those _any_ time you were in the RHS of a less-than arithmetic comparison. 👍